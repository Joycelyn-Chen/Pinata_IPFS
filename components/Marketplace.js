import Navbar from "./Navbar";
import NFTTile from "./NFTTile";
import MarketplaceJSON from "../Marketplace.json";
import axios from "axios";
import { useEffect,useState } from "react";

export default function Marketplace() {

// Sample data used for testing purposes
const sampleData = [
    {
        "name": "NFT#1",
        "description": "Cute Dog",
        "website":"http://axieinfinity.io",
        "image":"https://gateway.pinata.cloud/ipfs/QmTcQR1wMsnxakNikzePPJee77NxDtuRBkzvWeHva9vmfq",
        "price":"0.03ETH",
        "currentlySelling":"True",
        "address":"0xe81Bf5A757CB4f7F82a2F23b1e59bE45c33c5b13",
    },
    {
        "name": "NFT#2",
        "description": "Fantastic Dog",
        "website":"http://axieinfinity.io",
        "image":"https://gateway.pinata.cloud/ipfs/QmdUXeAUvjiwmS3htSYECxrr6ewjF27B3FYhuLfF33W1Bq",
        "price":"0.03ETH",
        "currentlySelling":"True",
        "address":"0xe81Bf5A757C4f7F82a2F23b1e59bE45c33c5b13",
    },
    {
        "name": "NFT#3",
        "description": "Next Generation of Dog",
        "website":"http://axieinfinity.io",
        "image":"https://gateway.pinata.cloud/ipfs/QmbkEvueGNBbHduz3DAUX4vVzcYn8M4CVfitzQP1nw9tRG",
        "price":"0.03ETH",
        "currentlySelling":"True",
        "address":"0xe81Bf5A757C4f7F82a2F23b1e59bE45c33c5b13",
    },
];

// useState hooks to manage component state
const [data, updateData] = useState(sampleData);
const [dataFetched, updateFetched] = useState(false);

// useEffect hook to fetch NFT data from the contract when the component mounts
useEffect(()=>{
    if (!dataFetched)
    getAllNFTs();

},[dataFetched])

// function to fetch all the NFTs from the smart contract
async function getAllNFTs() {
    const ethers = require('ethers');
    const provider = new ethers.providers.Web3Provider(window.ethereum);
    const signer = provider.getSigner();

    // attach the NFT marketplace contract
    let contract = new ethers.Contract(MarketplaceJSON.address, MarketplaceJSON.abi, signer);
    // fetch all the NFTs listed in the contract
    let transaction = await contract.getAllNFTs();

    console.log(transaction);
    // Fetch all the details of every NFT from the contract and display
    const items = await Promise.all(transaction.map(async i => {
        const tokenURI = await contract.tokenURI(i.tokenId);
        // Fetch metadata from IPFS for each NFT
        let meta = await axios.get(tokenURI);
        meta = meta.data;

        // Format the price to human-readable format
        let price = ethers.utils.formatUnits(i.price.toString(), 'ether');
        // create a new object with all the required details
        let item = {
            price,
            tokenId: i.tokenId.toNumber(),
            seller: i.seller,
            owner: i.owner,
            image: meta.image,
            name: meta.name,
            description: meta.description,
        }
        return item;
    }))

    updateFetched(true);
    updateData(items);
}


return (
    <div>
        <Navbar></Navbar>
        <div className="flex flex-col place-items-center mt-20">
            <div className="md:text-xl font-bold text-white">
                Top NFTs
            </div>
            <div className="flex mt-5 flex-wrap max-w-screen-xl text-center">
                {data.map((value, index) => {
                    return <NFTTile data={value} key={index}></NFTTile>;
                })}
            </div>
        </div>            
    </div>
);

}