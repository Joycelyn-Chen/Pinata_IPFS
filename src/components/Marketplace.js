import Navbar from "./Navbar";
import NFTTile from "./NFTTile";
import MarketplaceJSON from "../Marketplace.json";
import axios from "axios";
import { useEffect, useState } from "react";
import { enqueueSnackbar, useSnackbar } from 'notistack';

const responsive = {
    desktop: {
        breakpoint: { max: 3000, min: 1024 },
        items: 3,
        slidesToSlide: 3 // optional, default to 1.
    },
    tablet: {
        breakpoint: { max: 1024, min: 464 },
        items: 2,
        slidesToSlide: 2 // optional, default to 1.
    },
    mobile: {
        breakpoint: { max: 464, min: 0 },
        items: 1,
        slidesToSlide: 1 // optional, default to 1.
    }
};

export default function Marketplace() {
    const sampleData = [
        {
            "name": "NFT#1",
            "description": "Cute Dog",
            "website": "http://axieinfinity.io",
            "image": "https://gateway.pinata.cloud/ipfs/QmTcQR1wMsnxakNikzePPJee77NxDtuRBkzvWeHva9vmfq",
            "price": "0.03ETH",
            "currentlySelling": "True",
            "address": "0xe81Bf5A757CB4f7F82a2F23b1e59bE45c33c5b13",
        },
        {
            "name": "NFT#2",
            "description": "Fantastic Dog",
            "website": "http://axieinfinity.io",
            "image": "https://gateway.pinata.cloud/ipfs/QmdUXeAUvjiwmS3htSYECxrr6ewjF27B3FYhuLfF33W1Bq",
            "price": "0.03ETH",
            "currentlySelling": "True",
            "address": "0xe81Bf5A757C4f7F82a2F23b1e59bE45c33c5b13",
        },
        {
            "name": "NFT#3",
            "description": "Next Generation of Dog",
            "website": "http://axieinfinity.io",
            "image": "https://gateway.pinata.cloud/ipfs/QmbkEvueGNBbHduz3DAUX4vVzcYn8M4CVfitzQP1nw9tRG",
            "price": "0.03ETH",
            "currentlySelling": "True",
            "address": "0xe81Bf5A757C4f7F82a2F23b1e59bE45c33c5b13",
        },
    ];
    const [data, updateData] = useState(sampleData);
    const [dataFetched, updateFetched] = useState(false);
    const { enqueueSnackbar, closeSnackbar } = useSnackbar();
    useEffect(() => {
        if (!dataFetched)
            getAllNFTs();

    }, [dataFetched])

    async function getAllNFTs() {
        const ethers = require('ethers');
        const provider = new ethers.providers.Web3Provider(window.ethereum);
        const signer = provider.getSigner();
        let contract = new ethers.Contract(MarketplaceJSON.address, MarketplaceJSON.abi, signer);
        let transaction = await contract.getAllNFTs();

        // Fetch all the details of every NFT from the contract and display
        const items = await Promise.all(transaction.map(async i => {

            const tokenURI = await contract.tokenURI(i.tokenId);
            let meta = await axios.get(tokenURI,{  headers: {
                'Accept': 'text/plain'
              }});
            meta = meta.data;

            let price = ethers.utils.formatUnits(i.price.toString(), 'ether');
            let item = {
                price,
                tokenId: i.tokenId.toNumber(),
                seller: i.seller,
                owner: i.owner,
                image: meta.image,
                name: meta.name,
                description: meta.description,
            }
            return item;
        })).catch(err => { enqueueSnackbar(err, { autoHideDuration: 1000 }) })



        updateFetched(true);
        updateData(items);
    }


    return (
        <div>
            <Navbar></Navbar>
            <div className="flex flex-col place-items-center mt-20">
                <div className="md:text-xl font-bold text-white">
                    MOST POPULAR PIXELS
                </div>

                <div className="flex mt-5 flex-wrap max-w-screen-xl text-center">

                    {data.map((value, index) => {
                        return <NFTTile data={value} key={index}></NFTTile>;
                    })}

                </div>
            </div>
        </div>
    );

};