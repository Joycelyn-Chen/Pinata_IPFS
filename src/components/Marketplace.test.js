import { render, screen } from '@testing-library/react';
import Marketplace from './Marketplace';
import { MemoryRouter, Route, Router } from "react-router-dom";
import { createMemoryHistory } from 'history';
import renderer from 'react-test-renderer'



jest.mock('axios', () => ({
  get: jest.fn(() =>
    Promise.resolve({
      data: {
        name: 'NFT#1',
        description: 'Cute Dog',
        image: 'https://gateway.pinata.cloud/ipfs/QmTcQR1wMsnxakNikzePPJee77NxDtuRBkzvWeHva9vmfq',
      },
    })
  ),
}));

describe('Marketplace', () => {
  let history = null;
  beforeEach(()=>{
    history = createMemoryHistory({ initialEntries: ["/"] });
  })
  test('renders title', () => {
    
     const component = renderer.create(
        <Router location={history.location} navigator={history}>
        <Marketplace />
      </Router>
    );
    let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
    
  });

});
