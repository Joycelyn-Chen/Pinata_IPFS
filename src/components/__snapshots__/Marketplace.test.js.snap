// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Marketplace renders title 1`] = `
<div>
  <div
    className="min-w-screen bg-blue-500 bg-opacity-20"
  >
    <nav
      className="w-screen"
      style={
        Object {
          "width": "100%",
        }
      }
    >
      <ul
        className="flex flex-wrap items-center justify-between py-3 bg-transparent text-white pr-5"
      >
        <li
          className="flex items-center ml-5 pb-2"
        >
          <a
            href="/"
            onClick={[Function]}
          >
            <img
              alt=""
              className="inline-block -mt-2 animate-pulse"
              height={70}
              src="app_logo.png"
              width={70}
            />
            <div
              className="inline-block font-bold text-xl ml-2 tracking-widest"
            >
              Crypto Emporium
            </div>
          </a>
        </li>
        <li
          className="w-full lg:w-auto"
        >
          <div
            className="flex flex-wrap justify-center lg:justify-end"
          >
            <ul
              className="lg:flex justify-between font-bold text-lg mr-10"
            >
              <li
                className="border-b-2 hover:pb-0 p-2"
              >
                <a
                  href="/"
                  onClick={[Function]}
                >
                  Marketplace
                </a>
              </li>
              <li
                className="hover:border-b-2 hover:pb-0 p-2"
              >
                <a
                  href="/sellNFT"
                  onClick={[Function]}
                >
                  Add My NFT
                </a>
              </li>
              <li
                className="hover:border-b-2 hover:pb-0 p-2"
              >
                <a
                  href="/profile"
                  onClick={[Function]}
                >
                  Profile
                </a>
              </li>
              <li>
                <button
                  className="enableEthereumButton bg-blue-500 hover:bg-blue-700 text-white font-bold py-2 px-4 rounded text-sm transition-all duration-300 ease-in-out bg-yellow-500 hover:bg-yellow-600"
                  onClick={[Function]}
                >
                  Connect Wallet
                </button>
              </li>
            </ul>
          </div>
        </li>
      </ul>
    </nav>
    <div
      className="flex justify-center lg:justify-end mr-10"
    >
      <div
        className="text-white font-bold text-sm text-center lg:text-right "
      >
        Not Connected. Please agin login to view NFTs
      </div>
    </div>
  </div>
  <div
    className="flex flex-col place-items-center mt-20"
  >
    <div
      className="md:text-xl font-bold text-white"
    >
      MOST POPULAR PIXELS
    </div>
    <div
      className="flex mt-5 flex-wrap max-w-screen-xl text-center"
    >
      <a
        className="block w-full max-w-xs mx-auto mb-8 bg-white shadow-md hover:shadow-lg rounded-lg overflow-hidden relative"
        href="/nftPage/undefined"
        onClick={[Function]}
        style={
          Object {
            "marginLeft": "1px",
          }
        }
      >
        <div
          className="relative"
        >
          <img
            alt="NFT#1"
            className="h-64 w-full object-cover"
            src="https://gateway.pinata.cloud/ipfs/QmTcQR1wMsnxakNikzePPJee77NxDtuRBkzvWeHva9vmfq"
          />
          <div
            className="absolute top-0 left-0 w-full h-full bg-black opacity-25 rounded-lg"
          />
          <div
            className="absolute bottom-0 left-0 w-full flex flex-col justify-end px-4 py-2 bg-gray-100 opacity-75  hover:bg-black  transition-colors duration-100 overflow-y-auto"
          >
            <h3
              className="text-gray-700 uppercase mb-2 hover:text-white"
            >
              NFT#1
            </h3>
            <span
              className="text-gray-500 hover:text-white"
            >
              $
              0.03ETH
            </span>
            <p
              className="text-gray-600 mb-4 hover:text-white"
            >
              Cute Dog
            </p>
          </div>
          <span
            className="absolute top-0 right-0 bg-yellow-500 bg-opacity-75 text-white font-bold py-2 px-4 rounded-bl"
          >
            0.03ETH
          </span>
        </div>
      </a>
      <a
        className="block w-full max-w-xs mx-auto mb-8 bg-white shadow-md hover:shadow-lg rounded-lg overflow-hidden relative"
        href="/nftPage/undefined"
        onClick={[Function]}
        style={
          Object {
            "marginLeft": "1px",
          }
        }
      >
        <div
          className="relative"
        >
          <img
            alt="NFT#2"
            className="h-64 w-full object-cover"
            src="https://gateway.pinata.cloud/ipfs/QmdUXeAUvjiwmS3htSYECxrr6ewjF27B3FYhuLfF33W1Bq"
          />
          <div
            className="absolute top-0 left-0 w-full h-full bg-black opacity-25 rounded-lg"
          />
          <div
            className="absolute bottom-0 left-0 w-full flex flex-col justify-end px-4 py-2 bg-gray-100 opacity-75  hover:bg-black  transition-colors duration-100 overflow-y-auto"
          >
            <h3
              className="text-gray-700 uppercase mb-2 hover:text-white"
            >
              NFT#2
            </h3>
            <span
              className="text-gray-500 hover:text-white"
            >
              $
              0.03ETH
            </span>
            <p
              className="text-gray-600 mb-4 hover:text-white"
            >
              Fantastic Dog
            </p>
          </div>
          <span
            className="absolute top-0 right-0 bg-yellow-500 bg-opacity-75 text-white font-bold py-2 px-4 rounded-bl"
          >
            0.03ETH
          </span>
        </div>
      </a>
      <a
        className="block w-full max-w-xs mx-auto mb-8 bg-white shadow-md hover:shadow-lg rounded-lg overflow-hidden relative"
        href="/nftPage/undefined"
        onClick={[Function]}
        style={
          Object {
            "marginLeft": "1px",
          }
        }
      >
        <div
          className="relative"
        >
          <img
            alt="NFT#3"
            className="h-64 w-full object-cover"
            src="https://gateway.pinata.cloud/ipfs/QmbkEvueGNBbHduz3DAUX4vVzcYn8M4CVfitzQP1nw9tRG"
          />
          <div
            className="absolute top-0 left-0 w-full h-full bg-black opacity-25 rounded-lg"
          />
          <div
            className="absolute bottom-0 left-0 w-full flex flex-col justify-end px-4 py-2 bg-gray-100 opacity-75  hover:bg-black  transition-colors duration-100 overflow-y-auto"
          >
            <h3
              className="text-gray-700 uppercase mb-2 hover:text-white"
            >
              NFT#3
            </h3>
            <span
              className="text-gray-500 hover:text-white"
            >
              $
              0.03ETH
            </span>
            <p
              className="text-gray-600 mb-4 hover:text-white"
            >
              Next Generation of Dog
            </p>
          </div>
          <span
            className="absolute top-0 right-0 bg-yellow-500 bg-opacity-75 text-white font-bold py-2 px-4 rounded-bl"
          >
            0.03ETH
          </span>
        </div>
      </a>
    </div>
  </div>
</div>
`;
